### PR TITLE
feat: replace ClientOnly component with clientOnly function #82

### PR DESCRIPTION
**Pull request by pdanpdan**
Improves DX

### CODE REVIEW

**Comment by brillout** [comment]
I made some substantial changes, let me know what you think! See commit messages.

**Comment by pdanpdan** [comment]
Fixed one little glitch, the rest is ok with me.
We need to decide what to keep in the examples - should I clean up examples to original and just replace `ClientOnly` usage with `clienOnly`?

**Comment by brillout** [comment]
Ah, thanks for the fix.

Also, have you seen the comment [here](https://github.com/vikejs/vike-vue/pull/110#discussion_r1629802694) and [there](https://github.com/vikejs/vike-vue/pull/110#discussion_r1629804431)?

About the example, yea we should make it easier for the user to digest. We can showcase only one, or more than one interesting use case. It's /examples/**full** after all :)

**Comment by brillout** [comment]
Btw. I didn't test my changes, so maybe it doesn't work.

File: packages/vike-vue/src/utils/clientOnly.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+      onBeforeMount(() => {
+        const loader = source instanceof Function ? source : () => source
+        Promise.resolve(loader())
+          .then((component) => {
+            resolvedComp.value = 'default' in component ? component.default : component
+          })
+          .catch((e) => {
+            console.error('Failed loading component:', e)
+            throw e
```
[the remark this answers is not in the document]
Comment (pdanpdan) [reply packages/vike-vue/src/utils/clientOnly.ts:25]: The page breaks in a rather spectacular way :D
I'll go with logging error and swallowing it

[the remark this answers is not in the document]
Comment (brillout) [reply packages/vike-vue/src/utils/clientOnly.ts:25]: Ok. I guess Vue doesn't have "error boundaries" then (that's one thing I like about React).

[the remark this answers is not in the document]
Comment (pdanpdan) [reply packages/vike-vue/src/utils/clientOnly.ts:25]: I think I come up with something nicer - I'm also exposing the error to the fallback slot

**Comment by pdanpdan** [comment]
I think it's ready (including examples)

**Comment by brillout** [comment]
Very neat, `clientOnly.ts` looks really nice :sparkles:

There are quite a lot of examples, which I actually like :+1:. It's a dedicated page so it makes sense. It's nice for users to have a playground with lots of exampels to play with and going deep, while `vike.dev` can have easy docs while pointing to `/examples/full/`.

One last thing: the `utils/` directory is for small general purpose utilities like lodash. I think we should create a new directory for `clientOnly.ts`. Ideas for how to name that directory? Maybe `helpers/`? Maybe there is a terminology in Vue for these kind of helpers?

(I made some minor polishing. As always feel free to disagree and I'm happy if vike-vue eventually adopts a different style in the future.)

**Comment by pdanpdan** [comment]
I cannot find another good category name for this kind of functions.
The closest I can think of is `renderer` (it's a function that implements how something is rendered) / `render` (if we don't want to clutter `renderer`).
`wrappers` might work, or as a catchall `helpers`

File: packages/vike-vue/src/utils/clientOnly.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        // if the user doesn't want clientOnly() to use <template #fallback> then he should define a (empty) <template #client-only-fallback>
+        if (slots['fallback']) {
+          return slots['fallback']({ error: error.value, attrs })
+        }
+      }
+    },
+
+    slots: {} as SlotsType<{
+      fallback: { error: unknown; attrs: Record<string, any> }
```
Comment (pdanpdan) [inline packages/vike-vue/src/utils/clientOnly.ts:44]: Can we somehow make `attrs` get the inferred type and not a generic one?

Comment (brillout) [reply packages/vike-vue/src/utils/clientOnly.ts:44]: Good idea, that would be neat indeed :eyes:

**Comment by pdanpdan** [comment]
I'll move it to `helpers` - I realized it's publicly from the root of the package so it does not matter to much

**Comment by pdanpdan** [comment]
I cannot find a way to infer the `attrs` type, so until someone else figures it out I would say this should be ready

**Comment by brillout** [comment]
:+1: The PR LGTM.

@4350pChris WDTY?

**Review by 4350pChris — APPROVED** [review]

**Comment by 4350pChris** [comment]
@pdanpdan Would you be so kind and fix the merge conflicts?

**Comment by pdanpdan** [comment]
done

**Pull request merged**